### PR TITLE
added ANCE re-ranker

### LINKED ISSUE #3: ANCE as a re-ranker
From what I can tell, there's currently no way to use ANCE as a plain old re-ranker. While this is not the intention for ANCE in practice, it would be nice to have in some situations.

Would you accept a PR that adds an `ANCEReRanker`?

### PR DESCRIPTION
**Pull request by seanmacavaney**
closes #3

### CODE REVIEW

File: pyterrier_ance/__init__.py [hunk 1]
```diff
@@ -207,3 +207,63 @@ def _calc_scores(self,
...
+        config, tokenizer, model = load_model(self.args, checkpoint_path)
+        self.model = model
+        self.tokenizer = tokenizer
+        self.text_field = text_field
+
+    def __str__(self):
+        return "ANCE"
+
+    def transform(self, df):
+        from pyterrier import tqdm
```
Comment (cmacdonald) [inline pyterrier_ance/__init__.py]: dont think this is used.

File: pyterrier_ance/__init__.py [hunk 2]
```diff
@@ -207,3 +207,63 @@ def _calc_scores(self,
...
+        queries=[]
+        docs = []
+        for q in df["query"].to_list():
```
Comment (cmacdonald) [inline pyterrier_ance/__init__.py:245]: should you group by qid?

File: pyterrier_ance/__init__.py [hunk 3]
```diff
@@ -207,3 +207,63 @@ def _calc_scores(self,
                     rtr.append([query_id, pred_pid, docno, rank, scores[i]])
                     seen_pid.add(pred_pid)
         return pd.DataFrame(rtr, columns=["qid", "docid", "docno", "rank", "score"])
+
+
+class ANCEReRanker(TransformerBase):
```
Comment (cmacdonald) [inline pyterrier_ance/__init__.py]: I would describe this as a text scorer. How about ANCETextScorer?

File: pyterrier_ance/__init__.py (hunk 2 above)
Comment (cmacdonald) [reply pyterrier_ance/__init__.py:245]: alternatively, a pt.apply.by_query() function could be used?

**Review by cmacdonald — COMMENTED** [review]
a few comments. I think this is a text scoring re-ranker rather than an index reranker. Can we name it as such. And a few other nitpicks.

**Comment by seanmacavaney** [comment]
Hold on there! I wasn't ready for a review yet. Pushed to test on Colab.

**Comment by seanmacavaney** [comment]
@cmacdonald alright- ready now. May want to squash it upon merge because there was a bit of debugging.

groupby and pt.apply.by_query were not ideal because the ance package does some logging internally, which spams the output when done by individual query. (Also allows batching over documents across queries.)

**Pull request merged**